sql/api/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala

```diff
@@ -139,7 +139,9 @@ object JavaTypeInference {
       encoderFor(typeVariables(tv), seenTypeSet, typeVariables)
 
     case pt: ParameterizedType =>
-      encoderFor(pt.getRawType, seenTypeSet, JavaTypeUtils.getTypeArguments(pt).asScala.toMap)
+      val newTvs = JavaTypeUtils.getTypeArguments(pt).asScala.toMap
```

Comment:
For reference, a walkthrough:

```
StringFooWrapper extends FooWrapper<String>
FooWrapper<U> contains Foo<U>
Foo<T>
```

On the first iteration for StringFooWrapper, typeVariable map contains (U => String)
On the second iteration for FooWrapper, this gets overriden by (T => U)

The correct behavior is to keep both (U => String) and (T => U) in the map.